File: 0_1.baekjoon/dp/java/src/BOJ14728.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ14728 {
    private static int N, T;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        T = Integer.parseInt(st.nextToken());

        int[] dp = new int[T + 1];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            int k = Integer.parseInt(st.nextToken());
            int s = Integer.parseInt(st.nextToken());

            for (int j = T; j >= k; j--) {
                dp[j] = Math.max(dp[j], dp[j - k] + s);
            }
        }
        System.out.println(dp[T]);
    }
}

File: 0_1.baekjoon/dp/kotlin/src/BOJ14728.kt
import kotlin.math.max

fun main() = with(System.`in`.bufferedReader()) {
    val (n, t) = readLine().split(" ").map { it.toInt() }
    val dp = Array(n + 1) { IntArray(t + 1)}
    val times = IntArray(n + 1)
    val scores = IntArray(n + 1)

    for (i in 1..n) {
        val (k, s) = readLine().split(" ").map { it.toInt() }
        times[i] = k
        scores[i] = s
    }

    for (i in 1..n) {
        for (j in 1..t) {
            if (j - times[i] < 0) {
                dp[i][j] = dp[i-1][j]
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - times[i]] + scores[i])

            }
        }
    }

    println(dp[n][t])

}